plugins/fluent-bit/src/plugin.cpp

```diff
@@ -0,0 +1,608 @@
+
+/// A map of key-value pairs of Fluent Bit plugin configuration options.
+using property_map = std::map<std::string, std::string>;
+
+/// The arguments passed to the operator.
+struct operator_args {
+  std::string plugin;                           ///< Fluent Bit plugin name.
+  std::chrono::milliseconds poll_interval{250}; ///< Engine poll interval.
+  property_map service_properties;              ///< The global service options.
+  property_map args;                            ///< The plugin arguments.
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, operator_args& x) -> bool {
+    return f.object(x)
+      .pretty_name("operator_args")
+      .fields(f.field("plugin", x.plugin),
+              f.field("poll_interval", x.poll_interval),
+              f.field("service_properties", x.service_properties),
+              f.field("args", x.args));
+  }
+};
+
+/// A RAII-style wrapper around the Fluent Bit engine.
+class engine {
+  /// Callback that the Fluent Bit `lib` output invokes per record. We use when
+  /// the engine acts as source. Since we don't want to do any memory
+  /// management within Fluent Bit, we just make a copy of the data into our
+  /// shared buffer that we then process later with the source operator.
+  static auto handle_lib_output(void* record, size_t size, void* data) -> int {
+    auto str = std::string_view{reinterpret_cast<char*>(record), size};
+    auto* self = reinterpret_cast<engine*>(data);
+    self->append(str);
+    flb_lib_free(record);
+    return 0;
+  }
+
+public:
+  /// Constructs a Fluent Bit engine for use as "source" in a pipeline.
+  static auto
+  make_source(const operator_args& args, const record& plugin_config)
+    -> std::unique_ptr<engine> {
+    auto result
+      = make_engine(plugin_config, args.poll_interval, args.service_properties);
+    if (not result)
+      return nullptr;
+    if (not result->input(args.plugin, args.args))
+      return nullptr;
+    auto callback = flb_lib_out_cb{
+      .cb = handle_lib_output,
+      .data = result.get(),
+    };
+    // There are two options for the `lib` output:
+    // - format: "msgpack" or "json"
+    // - max_records: integer representing the maximum number of records to
+    //   process per single flush call.
+    if (not result->output("lib", {{"format", "json"}}, &callback))
+      return nullptr;
+    if (not result->start())
+      return nullptr;
+    return result;
+  }
+
+  /// Constructs a Fluent Bit engine for use as "sink" in a pipeline.
+  static auto make_sink(const operator_args& args, const record& plugin_config)
+    -> std::unique_ptr<engine> {
+    auto result
+      = make_engine(plugin_config, args.poll_interval, args.service_properties);
+    if (not result)
+      return nullptr;
+    if (not result->input("lib"))
+      return nullptr;
+    if (not result->output(args.plugin, args.args))
+      return nullptr;
+    if (not result->start())
+      return nullptr;
+    return result;
```

Comment:
Mind taking a look at e600c91ccc9881948e0745e40e8af942039ce2dc? This exposes the errors better.

Can we pair on the leak?

Reply:
https://github.com/tenzir/tenzir/commit/e600c91ccc9881948e0745e40e8af942039ce2dc improved the error display.

We paired on the leak.
The aforementioned leak seems to be caused by the AWS SDK _again_ because I was using `fluent-bit s3` as a sink. The AWS SDK had a memory leak problem that I already encountered when implementing the `s3` connector.